lib/mauth/request_and_response.rb

```diff
@@ -57,17 +149,36 @@ def signature_token
     def signature
       signature_info[:signature]
     end
+
+    def protocol_version
+      if !mcc_authentication.to_s.strip.empty?
+        2
+      elsif !x_mws_authentication.to_s.strip.empty?
+        1
+      end
+    end
```

Comment:
would it be better to just examine `signature_token` for this function? i.e. if it's v1 token => 1, if v2 => 2, else nil?

Reply:
@ejinotti-mdsol  that's how I had it at first but we would lose some clarity in our logging (or the `authenticate!` conditional would have to get more complicated).
For example if we had a req with the header `X-MWS-Authentication` (v1) and value `'MWSV2 123;signature'` with the V2 token. Then we would try authenticating with v2 and fail because of the mismatch in signature but I think we should be authenticating with V1 and failing because of the wrong token.

Reply:
Ah I see, header key mismatched with the token provided in the value.. hmmmmmmmm

Is it a requirement of the story that we have all of these various error cases to tell them exactly what they did wrong?  Like gave v1 key with v2 value, v2 key with v1 value, missing header entirely, etc?

Or could we instead encapsulate all of that checking here in this function like:
```ruby
if has_v1_key? && valid_v1_header_value?
  1
elsif has_v2_key? && valid_v2_header_value?
  2
else
  nil # and somewhere down the line this result in a generic you-did-something-wrong-lol error response?
end
```

Reply:
@ejinotti-mdsol  We have a different informative logging messages for all these scenarios but to an outside consumer any of these cases look identical (inauthentic), except for the case when an incoming request is a V1 (`X-MWS-Authentication`) header present, no v2 header present, and the `v2_authenticate_only` flag is true. (in this case the rack middleware will respond with an missing v2 error). 

The other complexity is that when the headers are missing altogether we want to raise `MAuthNotPresent` and log a warning instead of an error so that we don't spam error logs for services that deliberately don't include the header. (This was the subject of #23). 

The old authentication flow was: 
1) check for the presence of the header, log and raise if it's not there
1) check for the presence and validity of the time header, log and raise if it's not there 
1) check validity of the token, log and raise if it's not there
1) check the validity of the signature, log and raise if it's not there 

I tried to stay close to this by making the presence of the v1 or v2 header the outer conditional in `#authenticate`


We could encapsulate all of that checking here in this `protocol_versions` function as you said but I worry that would make the logging less clear for these cases of mismatched token and header name, which may be a pain point during the conversion process.

Reply:
Hmm, ok yes I guess there is not much more to do then.  Thanks!